lang/py/src/avro/io.py

```diff
@@ -341,6 +389,71 @@ def write_double(self, datum):
     self.write(chr((bits >> 48) & 0xFF))
     self.write(chr((bits >> 56) & 0xFF))
 
+  def write_decimal_bytes(self, datum, scale):
+    """
+    Decimal in bytes are encoded as long. Since size of packed value in bytes for
+    signed long is 8, 8 bytes are written.
+    """
+    sign, digits, exp = datum.as_tuple()
+    if exp > scale:
+      raise AvroTypeException('Scale provided in schema does not match the decimal')
+
+    unscaled_datum = 0
+    for digit in digits:
+      unscaled_datum = (unscaled_datum * 10) + digit
+
+    bits_req = unscaled_datum.bit_length()
+    if sign:
```

Comment:
It still seems like a few edge cases aren't correctly handled (e.g. `-128`). In case it helps, I copied a few tests below (the commented ones are currently failing).

```python
  def test_decimal_bytes_write(self):
    writers_schema = schema.parse("""{
      "type": "bytes",
      "logicalType": "decimal",
      "precision": 3,
      "scale": 2
    }""")

    writer, _, _ = write_datum(Decimal('2.56'), writers_schema)
    self.assertEquals(writer.getvalue(), b'\x04\x01\x00')

    writer, _, _ = write_datum(Decimal('-2.56'), writers_schema)
    self.assertEquals(writer.getvalue(), b'\x04\xff\x00')

    writer, _, _ = write_datum(Decimal('1.28'), writers_schema)
    self.assertEquals(writer.getvalue(), b'\x04\x00\x80')

    writer, _, _ = write_datum(Decimal('1.27'), writers_schema)
    self.assertEquals(writer.getvalue(), b'\x02\x7f')

    writer, _, _ = write_datum(Decimal('-1.28'), writers_schema)
    # self.assertEquals(writer.getvalue(), b'\x02\x80') # \x04\xff\x80

    writer, _, _ = write_datum(Decimal('1.00'), writers_schema)
    self.assertEquals(writer.getvalue(), b'\x02\x64')

    writer, _, _ = write_datum(Decimal('1'), writers_schema)
    # self.assertEquals(writer.getvalue(), b'\x02\x64') # \x02\x01

    # For consistency, it would also be great if the following passed:
    # with self.assertRaises(Exception):
    #   write_datum(Decimal('2.345'), writers_schema)
```